msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java

```diff
@@ -66,66 +107,270 @@ private void initializeSharedPreferenceFileManager(@NonNull final Context contex
+
+                                checkCurrentAccountNotifyCallback(callback, null);
+                            }
+                        }
+
+                        @Override
+                        public void onError(final Exception exception) {
+                            com.microsoft.identity.common.internal.logging.Logger.error(
+                                    TAG + methodName,
+                                    exception.getMessage(),
+                                    exception
+                            );
+                            callback.onError(exception);
+                        }
+                    }
+
+
+            );
+
+            ApiDispatcher.getAccounts(command);
+
+        }catch(MsalClientException clientException){
+            callback.onError(clientException);
+        }
+    }
+
+
+    private void getCurrentAccountFromSharedDevice(final CurrentAccountCallback callback, final PublicClientApplicationConfiguration configuration){
         new BrokerMsalController().getCurrentAccount(
                 configuration,
                 new TaskCompletedCallbackWithError<List<ICacheRecord>, Exception>() {
                     @Override
                     public void onTaskCompleted(List<ICacheRecord> cacheRecords) {
-                        MultiTenantAccount currentAccount = getPersistedCurrentAccount();
-                        MultiTenantAccount newAccount = getAccountFromICacheRecordList(cacheRecords);
-
-                        String currentAccountHomeAccountId = currentAccount == null ? "" : currentAccount.getHomeAccountId();
-                        String newAccountHomeAccountId = newAccount == null ? "" : newAccount.getHomeAccountId();
-
-                        boolean isCurrentAccountChanged = !currentAccountHomeAccountId.equalsIgnoreCase(newAccountHomeAccountId);
-                        if (isCurrentAccountChanged) {
-                            callback.onAccountChanged(currentAccount, newAccount);
-                        }
-
-                        persistCurrentAccount(cacheRecords);
-                        callback.onAccountLoaded(newAccount);
+                        checkCurrentAccountNotifyCallback(callback, cacheRecords);
                     }
 
                     @Override
                     public void onError(Exception exception) {
                         callback.onError(exception);
+
                     }
                 });
     }
 
+    private void checkCurrentAccountNotifyCallback(final CurrentAccountCallback callback, List<ICacheRecord> newAccountRecords){
+        IAccount localAccount = getPersistedCurrentAccount();
+        IAccount newAccount = newAccountRecords == null ? null : getAccountFromICacheRecordList(newAccountRecords);
+
+        if (localAccount == null) {
+            if (newAccount != null) {
+                callback.onAccountChanged(null, newAccount);
+            }
+        } else if (localAccount.getId() != newAccount.getId()) {
```

Comment:
This should be `!localAccount.getId().equals(newAccount.getId()))` since it's string comparison

Reply:
Fixed.